Stratis.Bitcoin/BlockPulling/LookaheadBlockPuller.cs

```diff
@@ -457,11 +529,27 @@ private Block NextBlockCore(CancellationToken cancellationToken)
                         // Or the block is still being downloaded or we need to ask for this block to be downloaded.
                         if (!isDownloading) AskBlocks(new ChainedBlock[] { header });
 
-                        OnStalling(header);
-                        this.IsStalling = true;
+                        this.OnStalling(header);
                     }
 
-                    WaitHandle.WaitAny(new[] { this.pushed, cancellationToken.WaitHandle }, WaitNextBlockRoundTimeMs);
+                    while (!cancellationToken.IsCancellationRequested)
```

Comment:
The weirdest part is that in the same method - NBC - you set this.consumed and you also wait for this.consumed AND you expect only one thread to exist in the node that executes NBC. That is the weird part. But due to the logic of NC that it actually calls `AskBlocks` and because `AskBlocks` then calls `ProcessQueue` it actually works. But is very nontrivial to grasp. The original idea with the thread would separate the logic and the design would be very clean, but yes, it would cost the thread :)